src/main/resources/assets/js/app/create/UserItemTypesTreeGrid.ts

```diff
@@ -121,51 +131,146 @@ export class UserItemTypesTreeGrid extends TreeGrid<UserTypeTreeGridItem> {
 
     setIdProvider(idProvider: IdProvider): void {
         this.idProviders = [idProvider];
-        this.manualidProvider = true;
+        this.manualIdProvider = true;
         this.addClass('flat');
     }
 
-    clearidProviders(): void {
+    reset(): void {
         this.idProviders = null;
-        this.manualidProvider = false;
+        this.manualIdProvider = false;
         this.removeClass('flat');
+        this.deselectAll(true);
+    }
+
+    protected createKeyBindings(): KeyBinding[] {
+        const result: KeyBinding[] = super.createKeyBindings();
+
+        result.push(new KeyBinding('tab', (event: KeyboardEvent) => {
+            this.focusNextOnTab(event);
+        }));
+
+        result.push(new KeyBinding('shift+tab', (event: KeyboardEvent) => {
+            this.focusPreviousOnShiftTab(event);
+        }));
+
+        return result;
+    }
+
+    private focusNextOnTab(event: KeyboardEvent): void {
+        if (this.isLastItemSelected()) {
+            this.focusElementOutOfGrid(event);
+        } else {
+            this.selectNextItemInGrid(event);
+        }
+    }
+
+    private isLastItemSelected(): boolean {
+        return this.getCurrentData().pop().getUserItem() === this.getFirstSelectedItem()?.getUserItem();
+    }
+
+    private focusElementOutOfGrid(event: KeyboardEvent): void {
+        event.preventDefault();
+        event.stopPropagation();
+        this.deselectAll(true);
+        this.elementToFocusWhenNoSelected?.giveFocus();
+    }
+
+    private selectNextItemInGrid(event: KeyboardEvent): void {
+        if (event.target !== this.elementToFocusWhenNoSelected.getHTMLElement()) {
+            event.preventDefault();
+            event.stopPropagation();
+        }
+
+        this.navigateDown();
     }
 
-    private initEventHandlers() {
+    private focusPreviousOnShiftTab(event: KeyboardEvent): void {
+        if (this.isFirstItemSelected()) {
+            this.focusElementOutOfGrid(event);
+        } else {
+            this.selectPreviousItemInGrid(event);
+        }
+    }
+
+    private isFirstItemSelected(): boolean {
+        return this.getCurrentData()[0].getUserItem() === this.getFirstSelectedItem()?.getUserItem();
+    }
+
+    private selectPreviousItemInGrid(event: KeyboardEvent): void {
+        if (event.target !== this.elementToFocusWhenNoSelected.getHTMLElement()) {
```

Comment:
Should be enough to check that `event.target` is inside the grid element (via `Element.contains()`).

Reply:
Updated